### DISCUSSION TITLE
Inline editing experience for relational interfaces

### DISCUSSION THREAD

**Discussion by hancewin** [directus/directus#17792] [Feature Requests]
### Summary

We'd love to have the opportunity to chose a different design for relations in the content editor.
More like strapis "dynamic zones" which just places the content of a relation directly into the UI.

For our non-code-experience customers the UX is confusing when you've to edit all the content in a separate modal box, it gets more confusing when you deep dive in the structure and have to trigger a modal in a modal and so on.

### Motivation

It help's non coder to get a more structured look in a M2A(Builder) relation case.

### Detailed Design

The Design could be very simple, just place the content at least for the first modal directly into the UI. Afterwards it could be opened in a modal.

### Must Have:
- Alternativ view for M2A relations

### Could Have:
- A decision for which one to use

### Drawbacks

should be just a small change

### Alternatives

-

### Adoption Strategy

I think it would be just a small UI Change which wouldn't need any further explanation

**Comment by pavelstianko** [comment]
I agree with you! Or at least create the item collapsable like this https://github.com/directus/directus/discussions/17402

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by rijkvanzanten** [comment]
Note to self: Adjust RFC to include any relational interface, not just M2A. Also ref https://github.com/directus/directus/discussions/17402

**Reply by w0kyj** [reply]
Although this is layout level inline editing, flagging these as related in functionality for inline editing: https://github.com/directus/directus/discussions/17937

**Comment by rijkvanzanten** [comment]
Heya! Thanks for the feature request 🙂  I think this makes a lot of sense to do for basically any of the relational interfaces.

Couple of additional questions that come to mind:

- How many "levels" of relationships should / can be inlined? 
- How are inlined fields visually presented? Should they be inset?
- How do we handle responsiveness for grid configurations in the related collection?
- Should the related items be regular "rows" (like they are now) that expand, or should all forms be visible inline all the time?
- If the form of a m2o is visible inline all the time, how do you select an existing item?
- What do you mean by _Afterwards it could be opened in a modal._ @hancewin? Do you mean that if you have nested relational fields they'd be shown in a modal?
- What happens if you have multiple levels of inset, do you have enough space for the nested fields?
  - Semi-related: if you nest an "inline" relational field _inside_ a "inline" relational field, how/when do we switch back to the modal view when you're "in too deep".

**Reply by pavelstianko** [reply]
What about this Rijk?
<img width="839" alt="Snímek obrazovky 2023-08-03 v 19 59 07" src="https://github.com/directus/directus/assets/13132639/e9dd487f-66f1-4d38-aef2-877539f2a6c8">

**Reply by hancewin** [reply]
Thank you for discussing this!

In my opinion, we require both a new inlined interface and the current modal interface to ensure a usable UI and UX that allows us to delve into as many relationships as necessary.

> How many "levels" of relationships should/can be inlined?

In my initial approach, I used our current Directus data and designed an m2a relationship with an m2m relationship within it:

![m2a in m2m](https://github.com/directus/directus/assets/40453910/567344ba-e7f5-45bb-82c8-8de3cc0eaaee)

Does this work? I believe it does. But does it provide a better overview? Perhaps. I think up to this point it's still relatively clear and could expedite the content editing.

So, how long could this work?

![Web 1920 – 2](https://github.com/directus/directus/assets/40453910/fd7038eb-fa81-4773-a905-e9b839c4bc19)

Does this work? Maybe. Does it improve the overview? Not at all. Therefore, in my opinion, the answer would be somewhere between the first and the second relationships. After that, it becomes more confusing than navigating through modals.

> If the form of an m2o is visible inline all the time, how do you select an existing item?

Perhaps, if you opt for the m2o to be inlined, we could have a small "select" button on the top bar of the relationship field. Once you choose one, you would see your fields filled out.

> What do you mean by "Afterwards it could be opened in a modal." @hancewin? Do you mean that if you have nested relational fields they'd be shown in a modal?

Yes, that's precisely what I meant.

> What happens if you have multiple levels of inset, do you have enough space for the nested fields?

As demonstrated, there should be a limit set that prohibits the display of relationships inlined beyond the x (imo the second) level.

**Reply by rijkvanzanten** [reply]
Thanks for your insight @hancewin! I'll have to think about a way to tell each "level" what "level" they're currently in in order for us to be able to dynamically switch between the inline and modal styles

**Reply by rijkvanzanten** [reply]
There was one more use case that just came to mind: I think it would also have to support drag and drop between different layers of the same recursive / nested interface. For example, if you have "page sections" with blocks:


```
- section-1
  - block-a
  - block-b
- section-2
  - block-c
  - block-d
```

I'd want to be able to drag `block-c` from `section-2` into `section-1`, even though they're technically fields in a different "item"

**Reply by LasseRosenow** [reply]
> There was one more use case that just came to mind: I think it would also have to support drag and drop between different layers of the same recursive / nested interface. For example, if you have "page sections" with blocks:
> 
> 
> ```
> - section-1
>   - block-a
>   - block-b
> - section-2
>   - block-c
>   - block-d
> ```
> 
> I'd want to be able to drag `block-c` from `section-2` into `section-1`, even though they're technically fields in a different "item"

I strongly second this! I am eagerly waiting for this feature so I could build a flexible component builder using rows, columns, containers and so on 😍😍.
Right now doing this is not ideal as the actual content would be hidden in some deeper layers below the rows and columns :)

**Comment by gastonbesada** [comment]
Hi! I'd like to have this new feature! I'd like to give my comments that I hope these can contribute to this topic.

From what I know of strapi, It has 2 components to manage this feature, Components and Dynamic zones.

Components - A group of Fields (relational collection), this can be repeatable.
https://docs.strapi.io/user-docs/content-type-builder/configuring-fields-content-type#-components
https://docs.strapi.io/user-docs/content-type-builder/creating-new-content-type#creating-a-new-component

Dynamic zones - A section in the Edit Item form that can have pre configured Components.
https://docs.strapi.io/user-docs/content-type-builder/configuring-fields-content-type#-dynamic-zones
https://docs.strapi.io/dev-docs/api/graphql#fetch-dynamic-zone-data

https://docs.strapi.io/dev-docs/api/entity-service/components-dynamic-zones

Here are some screenshots with the proposal.

Add Component Button in Data Model:
![screenshot0](https://github.com/directus/directus/assets/884940/b9a35879-03af-4e8b-beb2-3ef5a805ecc3)

New Component added to Data Model
![screenshot1](https://github.com/directus/directus/assets/884940/17e2d289-a117-4b4d-88df-a9b6cf3a8eb0)

Create Component  Setup
![screenshot3](https://github.com/directus/directus/assets/884940/f71e546f-8b07-487c-8097-7461563c915c)

Edit Component Fields and layout
![screenshot4](https://github.com/directus/directus/assets/884940/b29f1d1c-ac4e-4efb-ade6-76ee124ade54)

Add Component field to Parent Collection data model
![screenshot6](https://github.com/directus/directus/assets/884940/afcd8fe9-537b-425e-8601-7903ef9f929d)

Add Dynamic zone field to parent Collection data model
![screenshot8](https://github.com/directus/directus/assets/884940/dc8536b3-8458-4d19-a9e8-9158aad16e3b)

Component and Dynamic zone fields added to parent Collection
![screenshot13](https://github.com/directus/directus/assets/884940/42245457-7b6c-4f7c-a9c5-f2837d9cb5a1)
Add component to dyn zone must be a multi select tree pulldown with categories and its Collections, user can select many categories and/or collections

Create new parent item with Component and Dyn zone fields
![screenshot9](https://github.com/directus/directus/assets/884940/50f14432-8409-4d4a-9f11-c51b6b305dd0)

Add a components to the field 
![screenshot10](https://github.com/directus/directus/assets/884940/1725020f-d2c5-4313-b92b-41a38a77f966)
Items can be ordered doing drag and drop.

Add a Component to the dynamic zone section
![screenshot11](https://github.com/directus/directus/assets/884940/f1508d1f-1bfa-416b-a90e-b071476f52d7)

Edit the fields in the component in dynamic zone section
![screenshot12](https://github.com/directus/directus/assets/884940/22ba1dde-cfd6-4b0b-be77-18f3b97fd5f0)
Items can be ordered doing drag and drop.

Not sure if Group and Repeater field will still be necessary.

**Reply by rijkvanzanten** [reply]
As far as I can tell, it's "just" a one-to-many/many-to-many/many-to-any interface, but instead of opening the individual items in the drawer, they're rendered inline 🙂  (Although I'm sure my marketing friends would be very happy if we gave it a catchy name 🐒 )

**Reply by gastonbesada** [reply]
Yeah! Totally agree with you, it's something like an all-in-one inline relational edition plus grouping, nesting and optional repeating and ordering choices. This kind of features are very well received by content managers and some CMS have implemented it today. I think the idea for the name "Component" came from the concept of: a chunk of code or functionality being converted into "reusable" section. 

https://www.contentful.com/help/references/
https://hygraph.com/docs/editor-guides/content/field-types#components
https://hygraph.com/features/structure#components
https://docs.strapi.io/user-docs/content-manager/writing-content/#components

**Reply by pavelstianko** [reply]
Yes, I agree. It would be nice to have easy way how to create some component and assign it to selected collections. Like create component Seo (with fields og-title, og-desc, ...) and assign this component to all my singleton pages.

**Reply by pavelstianko** [reply]
Any updates on this one? I think it can make working with builder much easier and faster.

**Reply by marfal** [reply]
+1 Would also love to know how to accomplish that. I'm new to Directus and love the idea of quickly building my internal CRUD tools, but in most real-life scenarios you'll want to display relevant data subsets in the same panel, e.g. CRMs. So having this implemented as part of the internal tools rather than trying to build a front end for it would be great. Maybe there is a way for it already but I haven't seen it yet when checking the docs, YouTube, and most importantly my fav. 100 Apps in 100 Hours series :) :muscle:

**Comment by Pachat** [comment]
Is the team working on this very demanded feature?
Hygraph may give inspiration for the UI

**Reply by pavelstianko** [reply]
Hey Pachat, I found this amazing extension: https://github.com/smartlabsAT/directus-expandable-blocks#readme. So you can try this one!

**Reply by Pachat** [reply]
Very insteresting. For M2A.
Is it working (or easily customizable) with O2M Directus components? Or the overhead of using it as is for just 1 related collection is minimal?

**Reply by michelegranatiero** [reply]
Can it be used for many-to-many relationships? I believe M2M is the only missing piece for inline editing relationships

**Comment by michelegranatiero** [comment]
any updates on this? #17402

**Reply by JamesW1** [reply]
Hi @michelegranatiero - no updates, but if we pick this up you'll be able to follow on our public roadmap https://roadmap.directus.io/roadmap

In the meantime, these extensions provide some support for:
- M2A: https://github.com/smartlabsAT/directus-expandable-blocks#readme
- M2O: https://github.com/hanneskuettner/directus-extension-inline-form-interface

**Reply by michelegranatiero** [reply]
Hi @JamesW1 , I'm aware of those extensions but I would really need a custom extension for M2M, because when it comes to nested relations the drawers that appear on top of each other (stacked) with the "standard" interface are a little bit confusing for some users. Thank you for your answer, any suggestion is appreciated.

**Comment by michelegranatiero** [comment]
Hi, for those interested I built an interface extension for M2M relationships. It should give the possibility to nest inline items.

Search `inline m2m` in directus marketplace, or you can find the repo here: https://github.com/michelegranatiero/directus-extension-inline-m2m

Hope it can be useful!

**Reply by pavelstianko** [reply]
Wow, this looks amazing! 🤩
